src/render/src/encoder.rs

```diff
@@ -1,19 +1,101 @@
 //! Graphics commands encoder.
 
-#![deny(missing_docs)]
-
 use draw_state::target::{Depth, Stencil};
 use std::error::Error;
 use std::any::Any;
 use std::{fmt, mem};
+use std::sync::{mpsc, Arc, Mutex, MutexGuard};
 
-use core::{Backend, CommandQueue, GraphicsCommandPool, GraphicsQueue,
+use core::{Backend, CommandQueue, CommandPool,
            IndexType, SubmissionResult, VertexCount};
-use core::{self, command, format, handle, texture};
-use core::command::{GraphicsCommandBuffer, Submit};
-use core::memory::{self, cast_slice, Typed, Pod, Usage};
-use slice;
-use pso;
+use core::{self, command, format};
+use core::command::CommandBuffer;
+use memory::{self, cast_slice, Typed, Pod, Usage};
+use {handle, image};
+
+// This is the unique owner of the inner struct.
+pub struct Scope<B: Backend, C>(Arc<Mutex<ScopeInner<B, C>>>);
+pub struct ScopeGuard<'a, B: Backend, C: 'a> {
```

Comment:
Here is the `Mutex` version, I couldn't find a better way to write it. The API suffers a bit IMO.

Reply:
@kvark So do you think the internal safety is worth it here ?

Reply:
I would say no, unless we find a way to keep the same API.